### PR TITLE
Add configurable spooling filesystem layout implementations

### PR DESCRIPTION
**Pull request by wendigo**
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
## Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSegmentPruner.java [hunk 1]
```diff
@@ -95,14 +96,25 @@ long pruneExpiredBefore(Instant expiredBefore)
         if (closed) {
             return 0;
         }
+
+        long pruned = 0;
+        for (Location location : layout.searchPaths(rootLocation)) {
+            log.info("Pruning location: %s", location);
```
Comment (losipiuk) [inline plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSegmentPruner.java:102]: Probably it should not be info

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingConfig.java [hunk 2]
```diff
@@ -81,6 +83,19 @@ public FileSystemSpoolingConfig setLocation(String location)
...
+        return layout;
+    }
+
+    @Config("fs.layout")
+    @ConfigDescription("Spooled segments filesystem layout")
+    public FileSystemSpoolingConfig setLayout(Layout layout)
+    {
+        this.layout = layout;
+        return this;
+    }
```
Comment (losipiuk) [inline plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingConfig.java:97]: why to allow user to configure that?
Seems like it depends on what FS you use under the hood. If S3 - got with this layout. If GCS go with that.
Maybe MINIO vs S3 is the reason - but not sure if that justifies extra config option.

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java [hunk 3]
```diff
@@ -207,14 +210,8 @@ public SpooledSegmentHandle handle(SpooledLocation location)
         return new FileSystemSpooledSegmentHandle(encoding, queryId, uuid, Optional.of(encryptionHeadersTranslator.extractKey(location.headers())));
     }
 
-    private Location location(FileSystemSpooledSegmentHandle handle)
```
Comment (losipiuk) [inline plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java:210]: can you keep the method so call spots are clearer (you will not have to do `fileSystemLayout.` in call spots.

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingModule.java [hunk 4]
```diff
@@ -78,6 +82,19 @@ protected void setup(Binder binder)
...
+                fileSystemConfig -> fileSystemConfig.getLayout() == SIMPLE,
+                layoutBinder -> layoutBinder.bind(FileSystemLayout.class).to(SimpleFileSystemLayout.class)));
+
+        install(conditionalModule(
+                FileSystemSpoolingConfig.class,
+                fileSystemConfig -> fileSystemConfig.getLayout() == PARTITIONED,
+                layoutBinder -> {
+                    configBinder(layoutBinder).bindConfig(PartitionedLayoutConfig.class);
+                    layoutBinder.bind(FileSystemLayout.class).to(PartitionedFileSystemLayout.class);
+                }));
```
Comment (losipiuk) [inline plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingModule.java:97]: Like I said above - I think this should be based on FS type

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/PartitionedFileSystemLayout.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class PartitionedFileSystemLayout
+        extends SimpleFileSystemLayout
+{
+    private final int partitions;
+
+    public PartitionedFileSystemLayout(PartitionedLayoutConfig layoutConfig)
```
Comment (losipiuk) [inline plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/PartitionedFileSystemLayout.java:31]: do we really need that config?
I would just got with what is reasonable for S3. Probably you also want to denote that this layout is for S3 purposes in this name. It is not only partitioning which is important here, but how paths are built.

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/PartitionedLayoutConfig.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+package io.trino.spooling.filesystem;
+
+import io.airlift.configuration.Config;
+import io.airlift.configuration.ConfigDescription;
+import jakarta.validation.constraints.Max;
+import jakarta.validation.constraints.Min;
+
+public class PartitionedLayoutConfig
```
Comment (losipiuk) [inline plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/PartitionedLayoutConfig.java:21]: drop?

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/SimpleFileSystemLayout.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        return List.of(rootLocation);
+    }
+
+    public Optional<Instant> getExpiration(Location location)
+    {
+        String filename = location.fileName();
+        int index = filename.indexOf(SEGMENT_SEPARATOR);
+        if (index == -1) {
+            return Optional.empty(); // Not a segment
+        }
```
Comment (losipiuk) [inline plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/SimpleFileSystemLayout.java:46]: IllegalArgumentException?

File: plugin/trino-spooling-filesystem/src/test/java/io/trino/spooling/filesystem/TestFileSystemSegmentPruner.java [hunk 8]
```diff
@@ -41,110 +42,121 @@
...
+            .setLocation(TEST_LOCATION.toString())
             .setPruningBatchSize(1);
 
     @Test
     public void shouldPruneExpiredSegments()
     {
-        MemoryFileSystem fileSystem = new MemoryFileSystem();
-        try (ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor()) {
-            FileSystemSegmentPruner pruner = new FileSystemSegmentPruner(SPOOLING_CONFIG, _ -> fileSystem, executorService);
+        for (FileSystemLayout layout : LAYOUTS) {
```
Comment (losipiuk) [inline plugin/trino-spooling-filesystem/src/test/java/io/trino/spooling/filesystem/TestFileSystemSegmentPruner.java:57]: Create specific methods for each layout type.
Or if you prefer make this test abstract and create subclass per layout.
For loops in tests suck - and for some reason we do not want to have providers in code (I mean - I want - but not everyone)

**Review by losipiuk — APPROVED** [review]
Some suggestions

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingConfig.java (hunk 2 above)
Comment (wendigo) [reply plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingConfig.java:97]: I'd say that this is a deployment option

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSegmentPruner.java (hunk 1 above)
Comment (wendigo) [reply plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSegmentPruner.java:102]: Good point, fixed

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java (hunk 3 above)
Comment (wendigo) [reply plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java:210]: THe method will call the layout so it doesn't seem to make sense

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingModule.java (hunk 4 above)
Comment (wendigo) [reply plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingModule.java:97]: I disagree - for smaller deployments, you probably don't care about API quotas and the default (SIMPLE) is enough. For larger ones - you want to configure the amount of partitions

File: plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/SimpleFileSystemLayout.java (hunk 7 above)
Comment (wendigo) [reply plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/SimpleFileSystemLayout.java:46]: No, it should just skip a file

**Pull request merged**